Clean up warnings in library module

NthPortal opened the pull request:
Inspired by #8998

NthPortal committed:
Enable deprecation warnings for library build

Enable deprecation warnings for library module build
and fix or suppress all deprecations in the library module.

NthPortal committed:
Enable feature warnings for library build

Enable feature warnings for library module build
and import feature flags where required in the library module.

NthPortal committed:
Remove remaining -Xlint exclusions from library build

Remove remaining -Xlint exclusions from library build
and suppress locations where -Xlint raised warnings.

NthPortal commented:
I'm going to need to faff about with the ordering of the last 3 commits (and possibly squash 2 together) in order to get a clean history, I imagine

som-snytt commented:
This is the one Seth was looking forward to finishing up over a cold one? What a nice guy. He's probably off bumping an sbt version somewhere instead.

NthPortal committed:
Enable -Werror based on CI and for library build

Add `fatalWarnings` boolean setting key to sbt build
that adds the -Werror compile option to select modules.

Enable `fatalWarnings` setting for library module build.

Do not enable -Werror for the doc task.

NthPortal commented:
I'm a fork/join pool, and I steal work ;)

NthPortal commented on build.sbt:
```diff
@@ -424,6 +437,14 @@ val mimaFilterSettings = Seq {
+    ProblemFilters.exclude[MissingClassProblem]("scala.collection.immutable.HashMap$KeySet"),                 // private
+    ProblemFilters.exclude[MissingClassProblem]("scala.collection.immutable.HashMap$HashKeySet"),             // private
+    ProblemFilters.exclude[MissingTypesProblem]("scala.concurrent.impl.Promise$Transformation"),              // private[concurrent]
+
```
in a possibly vain attempt to avoid a merge conflict with #9232, I put the MiMa excludes for this PR above the MiMa excludes for the previous merged PR

NthPortal commented:
new error needs fixing/suppressing due to #8120

NthPortal commented:
Fix blocked by #9247 (or similar)

NthPortal commented on src/library/scala/math/BigInt.scala:
```diff
@@ -121,8 +123,8 @@ final class BigInt(val bigInteger: BigInteger)
-  /** Compares this BigInt with the specified value for equality.
-   */
+  /** Compares this BigInt with the specified value for equality. */
+  @nowarn("cat=other-non-cooperative-equals")
```
can't actually do this yet, as it doesn't compile pre-bootstreap. will have to filter on message until re-STARR

SethTisue reviewed (approved):
LGTM!

xuwei-k commented on src/library/scala/concurrent/impl/Promise.scala:
```diff
@@ -377,8 +377,9 @@ private[concurrent] object Promise {
     private[this] final val _xform: Int
-  ) extends DefaultPromise[T]() with Callbacks[F] with Runnable with Batchable with OnCompleteRunnable {
-    final def this(xform: Int, f: _ => _, ec: ExecutionContext) = this(f.asInstanceOf[Any => Any], ec.prepare(), null, xform)
+  ) extends DefaultPromise[T]() with Callbacks[F] with Runnable with Batchable {
+    final def this(xform: Int, f: _ => _, ec: ExecutionContext) =
```
https://github.com/akka/akka/issues/29980 🤔

NthPortal replied:
I'm concerned about how closely this is holding the standard library to implementation details

dwijnand replied:
It's not, IMO.  It's on Akka for relying on it.
